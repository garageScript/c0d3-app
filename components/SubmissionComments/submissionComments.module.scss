.wrapper {
  background-color: #fafbfc;
}

.text {
  padding-top: 15px;
  margin-right: 15px;
}
