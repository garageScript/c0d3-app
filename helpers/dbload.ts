import { Options, Sequelize } from 'sequelize'
import { Lesson, LessonTypes } from './models/Lesson'
import { User, UserTypes } from './models/User'
import { UserLessonTypes, UserLesson } from './models/UserLesson'
import { SubmissionTypes, Submission } from './models/Submission'
import { ChallengeTypes, Challenge } from './models/Challenge'
import { AlertTypes, Alert } from './models/Alert'
import { StarTypes, Star } from './models/Star'

const options: Options = {
  host: process.env.DB_HOST || 'city',
  logging: false,
  dialect: 'postgres',
  pool: {
    max: 5,
    min: 0,
    acquire: 30000,
    idle: 10000
  }
}

if (process.env.NODE_ENV != 'production') {
  options.port = ((process.env.DB_PORT as unknown) as number) || 5432
}
const sequelize = new Sequelize(
  process.env.DB_NAME || 'you',
  process.env.DB_USER || 'failed',
  process.env.DB_PW || 'this',
  options
)

Alert.init(AlertTypes, {
  tableName: 'alerts',
  sequelize
})

Lesson.init(LessonTypes, {
  tableName: 'lessons',
  sequelize
})

User.init(UserTypes, {
  tableName: 'users',
  sequelize
})

UserLesson.init(UserLessonTypes, {
  tableName: 'userLessons',
  sequelize
})

Submission.init(SubmissionTypes, {
  tableName: 'submissions',
  sequelize
})

Challenge.init(ChallengeTypes, {
  tableName: 'challenges',
  sequelize
})

Star.init(StarTypes, {
  tableName: 'stars',
  sequelize
})

Lesson.hasMany(Challenge, {
  as: 'challenges', // Defaults to Challenges
  foreignKey: 'lessonId'
})
Lesson.belongsToMany(User, {
  as: 'lesson',
  foreignKey: 'lessonId', // Defaults to LessonId
  through: { model: UserLesson }
})

Submission.belongsTo(User, { as: 'user' })
Submission.belongsTo(User, { as: 'reviewer' })
Submission.belongsTo(Challenge, { as: 'challenge' })
Submission.belongsTo(Lesson, { as: 'lesson' })

Star.belongsTo(User, { as: 'student' })
Star.belongsTo(User, { as: 'mentor' })
Star.belongsTo(Lesson, { as: 'lesson' })

User.belongsToMany(Lesson, {
  foreignKey: 'userId', // Defaults to UserId
  through: { model: UserLesson }
})

User.hasMany(Submission, {
  foreignKey: 'userId'
})

User.hasMany(UserLesson, {
  foreignKey: 'userId'
})

UserLesson.belongsTo(User, {
  foreignKey: 'userId',
  as: 'user'
})

sequelize.sync({ alter: !!process.env.ALTER_DB })

export default {
  Lesson,
  Challenge,
  Submission,
  User,
  UserLesson,
  Alert,
  Star,
  sequelize
}

export { User, UserLesson, options }
