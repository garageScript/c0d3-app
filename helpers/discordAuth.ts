import prisma from '../prisma'
import { URLSearchParams } from 'url'
import { User } from '.prisma/client'
import fetch from 'node-fetch'

const discordAPI = 'https://discordapp.com/api'
export const client_id = process.env.DISCORD_KEY
export const client_secret = process.env.DISCORD_SECRET
export const redirect_uri = process.env.DISCORD_REDIRECT_URI // {baseurl}/api/auth/callback/discord

type AccessTokenResponse = {
  access_token: string
  token_type: string
  expires_in: number
  refresh_token: string
  scope: string
}

type UserInfoResponse = {
  id: string
  username: string
  avatar: string
  discriminator: string
  public_flags: number
  locale: string
  mfa_enabled: boolean
  email: string
  verified: boolean
}

export type DiscordUserInfo = {
  userId: string
  username: string
  avatarUrl: string
  refreshToken: string
}

export const updateRefreshandAccessTokens = (
  userId: number,
  discordId: string,
  refreshToken: string,
  accessToken: string,
  accessTokenExpiresAt: Date
): Promise<User> => {
  return prisma.user.update({
    where: {
      id: userId
    },
    data: {
      discordRefreshToken: refreshToken,
      discordAccessToken: accessToken,
      discordAccessTokenExpires: accessTokenExpiresAt,
      discordId
    }
  })
}

const getTokenFromRefreshToken = (
  refresh_token: string
): Promise<AccessTokenResponse> => {
  return fetch(`${discordAPI}/oauth2/token`, {
    method: 'POST',
    headers: {
      'content-type': 'application/x-www-form-urlencoded;charset=utf-8'
    },
    body: new URLSearchParams({
      client_id,
      client_secret,
      grant_type: 'refresh_token',
      refresh_token
    })
  }).then(r => r.json())
}

const getUserInfoFromAccessToken = (
  accessToken: string
): Promise<UserInfoResponse> => {
  return fetch(`${discordAPI}/users/@me`, {
    method: 'GET',
    headers: {
      'content-type': 'application/x-www-form-urlencoded;charset=utf-8',
      Authorization: `Bearer ${accessToken}`
    }
  }).then(r => r.json())
}

export const getDiscordUserInfo = async (
  user: User
): Promise<DiscordUserInfo> => {
  try {
    const {
      discordRefreshToken,
      discordAccessToken,
      discordAccessTokenExpires
    } = user

    let userId = '',
      username = '',
      avatarUrl = '',
      refreshToken = discordRefreshToken || '',
      accessToken = discordAccessToken || ''

    if (!discordRefreshToken)
      return {
        userId,
        username,
        avatarUrl,
        refreshToken
      }

    if (!discordAccessToken || Number(discordAccessTokenExpires) < Date.now()) {
      const tokenResponse = await getTokenFromRefreshToken(refreshToken)
      refreshToken = tokenResponse.refresh_token || ''
      accessToken = tokenResponse.access_token || ''
      const accessTokenExpiresAt = new Date(
        Date.now() + (tokenResponse.expires_in * 1000 || 0)
      )

      // if updatedRefreshToken is undefined, empty string is stored in db to remove invalid tokens
      await updateRefreshandAccessTokens(
        user.id,
        user.discordId!,
        refreshToken,
        accessToken,
        accessTokenExpiresAt
      )
    }

    if (!accessToken)
      throw new Error(`refresh token invalid for userId ${user.id}`)

    const userInfo = await getUserInfoFromAccessToken(accessToken)

    userId = userInfo.id
    username = `${userInfo.username}#${userInfo.discriminator}`
    avatarUrl = `https://cdn.discordapp.com/avatars/${userInfo.id}/${userInfo.avatar}.png`

    if (!userId) throw new Error(`access token invalid for userId ${user.id}`)

    return {
      userId,
      username,
      avatarUrl,
      refreshToken
    }
  } catch (error) {
    return {
      userId: '',
      username: '',
      avatarUrl: '',
      refreshToken: ''
    }
  }
}
