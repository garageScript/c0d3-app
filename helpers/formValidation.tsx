import * as Yup from 'yup'

const TEXT_MIN = 2
const TEXT_MAX = 64
const PASSWORD_MIN = 6
const REGEX_ALPHANUMERICS_AND_SPACE = /^[a-zA-Z0-9_\s]*$/

const alertValidation = Yup.object({
  text: Yup.string().required('Required').strict(true),
  type: Yup.string().required('Required').strict(true),
  url: Yup.string(),
  urlCaption: Yup.string()
})

const lessonSchema = Yup.object({
  title: Yup.string().required('Required').strict(true),
  description: Yup.string().required('Required').strict(true),
  order: Yup.number()
    .required('Required')
    .typeError('Numbers only')
    .min(0, 'Positive numbers only')
    .strict(true)
})

const signupValidation = Yup.object({
  email: Yup.string().email('Invalid email address').required('Required'),
  username: Yup.string()
    .strict(true)
    .lowercase('Must be a lowercase string')
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required'),
  firstName: Yup.string()
    .strict(true)
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required'),
  lastName: Yup.string()
    .strict(true)
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required')
})

const loginValidation = Yup.object({
  username: Yup.string()
    .strict(true)
    .lowercase('Must be a lowercase string')
    .matches(REGEX_ALPHANUMERICS_AND_SPACE, 'Must be alphanumerics characters')
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .trim('Leading and trailing value must be alphanumeric character')
    .required('Required'),
  password: Yup.string()
    .min(PASSWORD_MIN, `Must be at least ${PASSWORD_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .required('Required')
})

const passwordValidation = Yup.object({
  password: Yup.string()
    .min(PASSWORD_MIN, `Must be at least ${PASSWORD_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .required('Required')
})

const confirmPasswordValidation = Yup.object({
  password: Yup.string()
    .min(PASSWORD_MIN, `Must be at least ${PASSWORD_MIN} characters`)
    .max(TEXT_MAX, `Must be ${TEXT_MAX} characters or less`)
    .required('Required'),
  confirmPassword: Yup.string()
    .oneOf([Yup.ref('password'), undefined], 'Password must match')
    .required('Required')
})

const resetPasswordValidation = Yup.object({
  userOrEmail: Yup.string()
    .min(TEXT_MIN, `Must be at least ${TEXT_MIN} characters`)
    .required('Required')
})

export {
  alertValidation,
  lessonSchema,
  signupValidation,
  loginValidation,
  passwordValidation,
  confirmPasswordValidation,
  resetPasswordValidation
}
