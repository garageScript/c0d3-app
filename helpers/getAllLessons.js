function importAll(r) {
  return r.keys().map(fileName => ({
    link: fileName.substr(1).replace(/\/index\.mdx$/, ''),
    module: r(fileName)
  }))
}

export const lessons = importAll(
  require.context('../pages/curriculum/lessons', true, /\.mdx$/)
)
