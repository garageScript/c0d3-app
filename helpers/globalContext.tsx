import React, { createContext, useState } from 'react'
import { GetAppQuery } from '../graphql'
import dummySession from '../__dummy__/sessionData'

//default context value is used for testing https://reactjs.org/docs/context.html#reactcreatecontext
export const GlobalContext = createContext<{
  session: GetAppQuery['session'] | null
  setContext: React.Dispatch<any>
}>({
  session: dummySession,
  setContext: () => {}
})

export const ContextProvider: React.FC = ({ children }) => {
  const [context, setContext] = useState<GetAppQuery['session'] | null>(null)
  return (
    <GlobalContext.Provider value={{ session: context, setContext }}>
      {children}
    </GlobalContext.Provider>
  )
}
