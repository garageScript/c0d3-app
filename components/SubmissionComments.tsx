import React, { useContext, useState } from 'react'
import { Comment, Submission } from '../graphql'
import Markdown from 'markdown-to-jsx'
import ReviewerProfile from './ReviewerProfile'
import { Button } from './theme/Button'
import { TrashIcon, PencilIcon } from '@primer/octicons-react'
import styles from '../scss/submissionComments.module.scss'
import { updateCache } from '../helpers/updateCache'
import { useDeleteCommentMutation, useEditCommentMutation } from '../graphql'
import { GlobalContext } from '../helpers/globalContext'
import { MdInput } from './MdInput'
import _ from 'lodash'
import Modal from 'react-bootstrap/Modal'

export const SubmissionComments: React.FC<{
  comments: Comment[]
  submission: Submission
}> = ({ comments, submission }) => {
  const [deleteComment] = useDeleteCommentMutation()
  const [editComment] = useEditCommentMutation()
  const [editing, setEditing] = useState<null | number>(null)
  const [currInput, setCurrInput] = useState('')
  const [show, setShow] = useState(false)
  const context = useContext(GlobalContext)
  const id = context.session?.user?.id

  const handleDeleteComment = (c: Comment) => {
    const update = updateCache({
      submissionId: submission.id,
      commentToDeleteId: c.id,
      lessonId: submission.lessonId,
      challengeId: submission.challengeId,
      userId: submission.user.id
    })

    deleteComment({
      variables: {
        id: c.id
      },
      update
    })
  }

  const handleStartEditing = (index: number, initialInput: string) => {
    if (editing !== null) setShow(true)
    else {
      setEditing(index)
      setCurrInput(initialInput)
    }
  }

  const handleUpdateComment = async (c: Comment, newContent: string) => {
    const update = updateCache({
      submissionId: submission.id,
      content: newContent,
      lessonId: submission.lessonId,
      challengeId: submission.challengeId,
      userId: submission.user.id
    })

    await editComment({
      variables: {
        id: c.id,
        content: newContent
      },
      update
    })

    setEditing(null)
    setCurrInput('')
  }

  const handleDiscardChange = () => {
    setEditing(null)
    setCurrInput('')
  }

  const handleModalClose = () => {
    setShow(false)
  }

  const modalText =
    'You can only edit one comment in a single comment chain at a time.'

  //to make sure comments load in correct order
  const submissionsComments = _.orderBy(comments, ['id'], ['asc'])

  return (
    <>
      <Modal show={show} onHide={handleModalClose}>
        <Modal.Header>
          <Modal.Title>Error</Modal.Title>
        </Modal.Header>
        <Modal.Body>{modalText}</Modal.Body>
        <Modal.Footer>
          <Button
            onClick={handleModalClose}
            type="info"
            size="sm"
            color="white"
          >
            Close
          </Button>
        </Modal.Footer>
      </Modal>
      {submissionsComments.map((c, i) => (
        <div
          key={`${c.content}${i}`}
          className="border border-lightgray rounded my-1"
        >
          <div className={'ms-3'}>
            {editing === i ? (
              <div className={styles['text']}>
                <MdInput
                  onChange={setCurrInput}
                  bgColor="white"
                  value={currInput}
                />
              </div>
            ) : (
              <Markdown wrapper="code_wrapper">{c.content}</Markdown>
            )}
          </div>
          <div
            className={`${styles['wrapper']} d-flex justify-content-between align-items-center p-1`}
          >
            <ReviewerProfile
              name={c.author!.name}
              username={c.author!.username}
              inline
            />
            {id === c.authorId ? (
              editing === i ? (
                <div>
                  <Button
                    type="light"
                    color="black"
                    onClick={handleDiscardChange}
                    data-testid="discard-button"
                  >
                    Discard Changes
                  </Button>
                  <span> </span>
                  <Button
                    type="info"
                    color="white"
                    onClick={() => {
                      handleUpdateComment(c, currInput)
                    }}
                    data-testid="save-button"
                  >
                    Save Changes
                  </Button>
                </div>
              ) : (
                <div>
                  <Button
                    outline
                    type="info"
                    size="sm"
                    color="info"
                    onClick={() => {
                      handleStartEditing(i, c.content)
                    }}
                    data-testid="edit-button"
                  >
                    <PencilIcon />
                  </Button>
                  <Button
                    outline
                    type="danger"
                    size="sm"
                    color="danger"
                    onClick={() => handleDeleteComment(c)}
                    data-testid="delete-button"
                  >
                    <TrashIcon />
                  </Button>
                </div>
              )
            ) : (
              <></>
            )}
          </div>
        </div>
      ))}
    </>
  )
}
