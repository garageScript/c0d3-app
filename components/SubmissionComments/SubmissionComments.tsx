import React, { useContext, useState } from 'react'
import { Comment, Submission } from '../../graphql'
import ReviewerProfile from '../ReviewerProfile'
import { Button } from '../theme/Button'
import { TrashIcon, PencilIcon } from '@primer/octicons-react'
import styles from './submissionComments.module.scss'
import { updateCache } from '../../helpers/updateCache'
import { useDeleteCommentMutation, useEditCommentMutation } from '../../graphql'
import { GlobalContext } from '../../helpers/globalContext'
import MdInput from '../MdInput'
import _ from 'lodash'
import Modal from 'react-bootstrap/Modal'
import {
  ModalBody,
  ModalFooter,
  ModalHeader,
  ModalTitle
} from 'react-bootstrap'
import HighlightMarkdown from '../mdx/HighlightMarkdown/HighlightMarkdown'

const SubmissionComments: React.FC<{
  comments: Comment[]
  submission: Submission
}> = ({ comments, submission }) => {
  const [deleteComment] = useDeleteCommentMutation()
  const [editComment] = useEditCommentMutation()
  const [editing, setEditing] = useState<null | number>(null)
  const [currInput, setCurrInput] = useState('')
  const [show, setShow] = useState(false)
  const context = useContext(GlobalContext)
  const id = context.session?.user?.id

  const handleDeleteComment = (c: Comment) => {
    const update = updateCache({
      submissionId: submission.id,
      commentToDeleteId: c.id,
      lessonId: submission.lessonId,
      challengeId: submission.challengeId,
      userId: submission.user.id
    })

    deleteComment({
      variables: {
        id: c.id
      },
      update
    })
  }

  const handleStartEditing = (index: number, initialInput: string) => {
    if (editing !== null) setShow(true)
    else {
      setEditing(index)
      setCurrInput(initialInput)
    }
  }

  const handleUpdateComment = async (c: Comment, newContent: string) => {
    const update = updateCache({
      submissionId: submission.id,
      content: newContent,
      lessonId: submission.lessonId,
      challengeId: submission.challengeId,
      userId: submission.user.id
    })

    await editComment({
      variables: {
        id: c.id,
        content: newContent
      },
      update
    })

    setEditing(null)
    setCurrInput('')
  }

  const handleDiscardChange = () => {
    setEditing(null)
    setCurrInput('')
  }

  const handleModalClose = () => {
    setShow(false)
  }

  const modalText =
    'You can only edit one comment in a single comment chain at a time.'

  //to make sure comments load in correct order
  const submissionsComments = _.orderBy(comments, ['id'], ['asc'])

  return (
    <>
      <Modal show={show} onHide={handleModalClose}>
        <ModalHeader>
          <ModalTitle>Error</ModalTitle>
        </ModalHeader>
        <ModalBody>{modalText}</ModalBody>
        <ModalFooter>
          <Button
            onClick={handleModalClose}
            btnType="info"
            size="sm"
            color="white"
          >
            Close
          </Button>
        </ModalFooter>
      </Modal>
      {submissionsComments.map((c, i) => (
        <div
          key={`${c.content}${i}`}
          className="border border-lightgray rounded my-1"
        >
          <div className="mx-3">
            {editing === i ? (
              <div className={styles['text']}>
                <MdInput
                  onChange={setCurrInput}
                  bgColor="white"
                  value={currInput}
                />
              </div>
            ) : (
              <div className="my-3">
                <HighlightMarkdown>{c.content}</HighlightMarkdown>
              </div>
            )}
          </div>
          <div
            className={`${styles.wrapper} d-flex justify-content-between align-items-center p-2 rounded`}
          >
            <ReviewerProfile
              name={c.author!.name}
              username={c.author!.username}
              inline
            />
            {id === c.authorId ? (
              editing === i ? (
                <div>
                  <Button
                    btnType="light"
                    color="black"
                    onClick={handleDiscardChange}
                    data-testid="discard-button"
                  >
                    Discard Changes
                  </Button>
                  <span> </span>
                  <Button
                    btnType="info"
                    color="white"
                    onClick={() => {
                      handleUpdateComment(c, currInput)
                    }}
                    data-testid="save-button"
                  >
                    Save Changes
                  </Button>
                </div>
              ) : (
                <div>
                  <Button
                    outline
                    btnType="info"
                    size="sm"
                    color="info"
                    onClick={() => {
                      handleStartEditing(i, c.content)
                    }}
                    data-testid="edit-button"
                  >
                    <PencilIcon />
                  </Button>
                  <Button
                    outline
                    btnType="danger"
                    size="sm"
                    color="danger"
                    onClick={() => handleDeleteComment(c)}
                    data-testid="delete-button"
                  >
                    <TrashIcon />
                  </Button>
                </div>
              )
            ) : (
              <></>
            )}
          </div>
        </div>
      ))}
    </>
  )
}

export default SubmissionComments
