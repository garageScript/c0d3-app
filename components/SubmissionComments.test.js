import React, { useContext, useEffect } from 'react'
import { SubmissionComments } from './SubmissionComments'
import { fireEvent, waitFor, render, screen } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import GET_PREVIOUS_SUBMISSIONS from '../graphql/queries/getPreviousSubmissions'
import { InMemoryCache } from '@apollo/client'
import submissionData from '../__dummy__/submission'
import DELETE_COMMENT from '../graphql/queries/deleteComment'
import EDIT_COMMENT from '../graphql/queries/editComment'
import dummySessionData from '../__dummy__/sessionData'
import { GlobalContext } from '../helpers/globalContext'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'

const submissionsData = [submissionData, { ...submissionData, id: 101 }]

const getPreviousSubmissionsMock = {
  request: {
    query: GET_PREVIOUS_SUBMISSIONS,
    variables: { challengeId: 6, userId: 3 }
  },
  result: {
    data: submissionsData
  }
}

const mocks = [
  getPreviousSubmissionsMock,
  {
    request: {
      query: DELETE_COMMENT,
      variables: { id: 1 }
    },
    result: { data: { id: 1 } },
    newData: jest.fn(() => ({
      data: {
        deleteComment: {
          id: '1'
        }
      }
    }))
  },
  {
    request: {
      query: EDIT_COMMENT,
      variables: { id: 1, content: 'edited test comment' }
    },
    result: { data: { id: 1, content: 'edited test comment' } },
    newData: jest.fn(() => ({
      data: {
        editComment: {
          id: '1',
          content: 'edited test comment'
        }
      }
    }))
  }
]

const comments = [
  {
    id: 1,
    content: 'test comment',
    submissionId: 101,
    createdAt: '124',
    authorId: 1,
    line: 1,
    fileName: 'js7/1.js',
    author: {
      username: 'admin'
    }
  },
  {
    id: 2,
    content: 'test comment 2',
    submissionId: 101,
    createdAt: '224',
    authorId: 1,
    line: 2,
    fileName: 'js7/1.js',
    author: {
      username: 'admin'
    }
  }
]

describe('Test SubmissionComments Component', () => {
  const fillOutCommentInput = async (getByTestId, str) => {
    const textField = getByTestId('textbox')

    // the type event needs to be delayed so the Formik validations finish
    await userEvent.type(textField, str, { delay: 1 })
  }

  const Wrapper = ({ userId = 1, children }) => {
    const context = useContext(GlobalContext)
    const incorrectUserData = {
      id: userId,
      name: 'fake user',
      email: 'fake@fakemail.com',
      isAdmin: true
    }
    useEffect(() => {
      context.setContext({ ...dummySessionData, user: incorrectUserData })
    }, [])
    return <>{children}</>
  }
  it('Should render correctly', () => {
    const { container } = render(
      <MockedProvider>
        <SubmissionComments comments={comments} submission={submissionData} />
      </MockedProvider>
    )

    expect(container).toMatchSnapshot()
  })

  it('Should delete comment when delete button is clicked', async () => {
    const cache = new InMemoryCache({ addTypename: false })

    cache.writeQuery({
      query: GET_PREVIOUS_SUBMISSIONS,
      variables: { userId: 3, challengeId: 6 },
      data: { getPreviousSubmissions: submissionsData }
    })
    const { container } = render(
      <MockedProvider cache={cache} addTypename={false} mocks={mocks}>
        <Wrapper>
          <SubmissionComments comments={comments} submission={submissionData} />
        </Wrapper>
      </MockedProvider>
    )

    const deleteButton = container.querySelector('.btn-outline-danger')
    fireEvent.click(deleteButton)

    const deleteCommentMutation = mocks[1].newData

    await waitFor(() => {
      expect(deleteCommentMutation).toHaveBeenCalled()
    })
  })

  it('should allow user to edit comment and discard changes when edit button is pressed, but not edit more than 1 comment in a chain', async () => {
    const cache = new InMemoryCache({ addTypename: false })

    cache.writeQuery({
      query: GET_PREVIOUS_SUBMISSIONS,
      variables: { userId: 3, challengeId: 6 },
      data: { getPreviousSubmissions: submissionsData }
    })
    expect.assertions(4)
    const { container, findByText } = render(
      <MockedProvider cache={cache} addTypename={false} mocks={mocks}>
        <Wrapper>
          <SubmissionComments comments={comments} submission={submissionData} />
        </Wrapper>
      </MockedProvider>
    )

    const editButton = container.querySelector('.btn-outline-info.btn-sm')
    fireEvent.click(editButton)

    await waitFor(() => {
      const discardButton = container.querySelector('.btn-light')
      expect(discardButton).toBeInTheDocument()
      expect(container.querySelector('.btn-info')).toBeInTheDocument()

      const editButtonTwo = container.querySelector('.btn-outline-info.btn-sm')
      fireEvent.click(editButtonTwo)

      const warningText =
        'You can only edit one comment in a single comment chain at a time.'

      waitFor(() => {
        expect(findByText(warningText)).toBeTruthy()
      })

      fireEvent.click(discardButton)
      waitFor(() => {
        expect(editButton).toBeInTheDocument()
      })
    })
  })
  it('should update comment when save changes button is clicked', async () => {
    const cache = new InMemoryCache({ addTypename: false })

    cache.writeQuery({
      query: GET_PREVIOUS_SUBMISSIONS,
      variables: { userId: 3, challengeId: 6 },
      data: { getPreviousSubmissions: submissionsData }
    })
    const { container, getByTestId } = render(
      <MockedProvider cache={cache} addTypename={false} mocks={mocks}>
        <Wrapper>
          <SubmissionComments
            comments={comments.slice(0, 1)}
            submission={submissionData}
          />
        </Wrapper>
      </MockedProvider>
    )

    await userEvent.click(container.querySelector('.btn-outline-info.btn-sm'))

    await userEvent.clear(getByTestId('textbox'))
    await fillOutCommentInput(getByTestId, 'edited test comment') // previous text: test comment

    const saveButton = container.querySelector('.btn-info')
    await userEvent.click(saveButton)

    const editCommentMutation = mocks[2].newData

    expect(editCommentMutation).toHaveBeenCalled()

    expect(container.querySelector('.btn-outline-danger')).toBeInTheDocument()
    expect(
      container.querySelector('.btn-outline-info.btn-sm')
    ).toBeInTheDocument()
  })

  it('should close Modal when you click Close button', async () => {
    const cache = new InMemoryCache({ addTypename: false })

    cache.writeQuery({
      query: GET_PREVIOUS_SUBMISSIONS,
      variables: { userId: 3, challengeId: 6 },
      data: { getPreviousSubmissions: submissionsData }
    })

    const { container } = render(
      <MockedProvider cache={cache} addTypename={false} mocks={mocks}>
        <Wrapper>
          <SubmissionComments comments={comments} submission={submissionData} />
        </Wrapper>
      </MockedProvider>
    )

    await userEvent.click(container.querySelector('.btn-outline-info.btn-sm'))
    await userEvent.click(container.querySelector('.btn-outline-info.btn-sm'))

    await waitFor(() => fireEvent.click(screen.getByText(/Close/)))
    await waitFor(() => expect(screen.queryByText('Error')).toBeFalsy())
  })
})
