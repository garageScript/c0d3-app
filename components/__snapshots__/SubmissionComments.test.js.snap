// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SubmissionComments Component Should render correctly 1`] = `
<div>
  <div
    class="border border-lightgray rounded my-1"
  >
    <div
      class="ms-3"
    >
      <span
        wrapper="code_wrapper"
      >
        test comment
      </span>
    </div>
    <div
      class="wrapper d-flex justify-content-between align-items-center p-1"
    >
      <div
        class="wrapper"
      >
        <a
          class="comment_author__inline mt-2 "
          href="/profile/admin"
        >
          <div
            class="d-inline"
          >
             
          </div>
          <div
            class="d-inline text-muted"
          >
             @admin
          </div>
        </a>
      </div>
      <div>
        <button
          class="btn borderless btn-outline-info onFocusBgFix btn-outline-bg-info btn-sm"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-pencil"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M11.013 1.427a1.75 1.75 0 012.474 0l1.086 1.086a1.75 1.75 0 010 2.474l-8.61 8.61c-.21.21-.47.364-.756.445l-3.251.93a.75.75 0 01-.927-.928l.929-3.25a1.75 1.75 0 01.445-.758l8.61-8.61zm1.414 1.06a.25.25 0 00-.354 0L10.811 3.75l1.439 1.44 1.263-1.263a.25.25 0 000-.354l-1.086-1.086zM11.189 6.25L9.75 4.81l-6.286 6.287a.25.25 0 00-.064.108l-.558 1.953 1.953-.558a.249.249 0 00.108-.064l6.286-6.286z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          class="btn borderless btn-outline-danger onFocusBgFix btn-outline-bg-danger btn-sm"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-trash"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M6.5 1.75a.25.25 0 01.25-.25h2.5a.25.25 0 01.25.25V3h-3V1.75zm4.5 0V3h2.25a.75.75 0 010 1.5H2.75a.75.75 0 010-1.5H5V1.75C5 .784 5.784 0 6.75 0h2.5C10.216 0 11 .784 11 1.75zM4.496 6.675a.75.75 0 10-1.492.15l.66 6.6A1.75 1.75 0 005.405 15h5.19c.9 0 1.652-.681 1.741-1.576l.66-6.6a.75.75 0 00-1.492-.149l-.66 6.6a.25.25 0 01-.249.225h-5.19a.25.25 0 01-.249-.225l-.66-6.6z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
  <div
    class="border border-lightgray rounded my-1"
  >
    <div
      class="ms-3"
    >
      <span
        wrapper="code_wrapper"
      >
        test comment 2
      </span>
    </div>
    <div
      class="wrapper d-flex justify-content-between align-items-center p-1"
    >
      <div
        class="wrapper"
      >
        <a
          class="comment_author__inline mt-2 "
          href="/profile/admin"
        >
          <div
            class="d-inline"
          >
             
          </div>
          <div
            class="d-inline text-muted"
          >
             @admin
          </div>
        </a>
      </div>
      <div>
        <button
          class="btn borderless btn-outline-info onFocusBgFix btn-outline-bg-info btn-sm"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-pencil"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M11.013 1.427a1.75 1.75 0 012.474 0l1.086 1.086a1.75 1.75 0 010 2.474l-8.61 8.61c-.21.21-.47.364-.756.445l-3.251.93a.75.75 0 01-.927-.928l.929-3.25a1.75 1.75 0 01.445-.758l8.61-8.61zm1.414 1.06a.25.25 0 00-.354 0L10.811 3.75l1.439 1.44 1.263-1.263a.25.25 0 000-.354l-1.086-1.086zM11.189 6.25L9.75 4.81l-6.286 6.287a.25.25 0 00-.064.108l-.558 1.953 1.953-.558a.249.249 0 00.108-.064l6.286-6.286z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <button
          class="btn borderless btn-outline-danger onFocusBgFix btn-outline-bg-danger btn-sm"
        >
          <svg
            aria-hidden="true"
            class="octicon octicon-trash"
            fill="currentColor"
            height="16"
            role="img"
            style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
            viewBox="0 0 16 16"
            width="16"
          >
            <path
              d="M6.5 1.75a.25.25 0 01.25-.25h2.5a.25.25 0 01.25.25V3h-3V1.75zm4.5 0V3h2.25a.75.75 0 010 1.5H2.75a.75.75 0 010-1.5H5V1.75C5 .784 5.784 0 6.75 0h2.5C10.216 0 11 .784 11 1.75zM4.496 6.675a.75.75 0 10-1.492.15l.66 6.6A1.75 1.75 0 005.405 15h5.19c.9 0 1.652-.681 1.741-1.576l.66-6.6a.75.75 0 00-1.492-.149l-.66 6.6a.25.25 0 01-.249.225h-5.19a.25.25 0 01-.249-.225l-.66-6.6z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
